File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let steps (number: int): int option =
    let rec collatz (number: int, steps: int): int option =
        match number with
        | 1 -> Some (steps)
        | x when x < 1 -> None
        | x when x % 2 = 0 -> collatz (x / 2, steps + 1)
        | x when x % 2 = 1 -> collatz (3 * x + 1, steps + 1)
        | _ -> None

    collatz(number, 0)

File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {

   if (start < 1) {
      return -1;
   }

   int number_of_steps = 0;
   
   while (start > 1) {
      if (start % 2 == 0) {
         start /= 2;
      } else {
         start = 3 * start + 1;
      }

      number_of_steps++;
   }

   return number_of_steps;
}
